Proofreading underlines may remain if Writing Tools session ends before animation finishes

lilyspiniolas opened the pull request:
#### d054860674a082f694c6fc443feda5c1cdb9a8e0
<pre>
Proofreading underlines may remain if Writing Tools session ends before animation finishes
<a href="https://bugs.webkit.org/show_bug.cgi?id=286083">https://bugs.webkit.org/show_bug.cgi?id=286083</a>
<a href="https://rdar.apple.com/142980582">rdar://142980582</a>

Reviewed by Richard Robinson.

Ensure the flush completion handler is invoked when the Writing Tools
session ends before the animation completes. As a result, the
WritingToolsController now receives a &apos;willEnd&apos; call, which removes
the underlines.

* Source/WebKit/WebKitSwift/WritingTools/WKIntelligenceReplacementTextEffectCoordinator.swift:
(WKIntelligenceReplacementTextEffectCoordinator.replacementEffectDidComplete(_:)):

Canonical link: <a href="https://commits.webkit.org/289073@main">https://commits.webkit.org/289073@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ff952a4156e9dabd9e17d8473a4d8de7e52dca8e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/85208 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/4943 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/39639 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/90335 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/36248 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/87297 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/5032 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/12918 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/66247 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/24061 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/88254 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/3843 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/77394 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/46515 "Found 2 new API test failures: /WPE/TestSSL:/webkit/WebKitWebView/web-socket-client-side-certificate, /WPE/TestSSL:/webkit/WebKitWebView/web-socket-tls-errors (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/3689 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/31655 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/35316 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/74452 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/32489 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/91753 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/12555 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/9142 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/74762 "Found 72 new test failures: compositing/blend-mode/non-separable-blend-modes.html compositing/layer-creation/will-change-layer-creation.html css3/filters/backdrop/backdrop-filter-does-not-size-properly-absolute.html editing/inserting/insert-list-user-select-none-crash.html editing/pasteboard/copy-paste-across-shadow-boundaries-with-style-2.html editing/spelling/grammar-and-spelling-error-styling.html fast/css-custom-paint/out-of-memory-while-adding-worklet-module.html fast/css/focus-ring-exists-for-search-field.html fast/forms/textarea-node-removed-from-document-crash.html fast/hidpi/filters-drop-shadow.html ... (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/12784 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/73231 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/73881 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18294 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/18297 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/16729 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/4542 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/12498 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/17950 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/12328 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/15821 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/14078 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebKit/WebKitSwift/WritingTools/WKIntelligenceReplacementTextEffectCoordinator.swift:
```diff
@@ -345,7 +345,12 @@ extension WKIntelligenceReplacementTextEffectCoordinator: PlatformIntelligenceTe
             self.replacementQueue.removeFirst()
-            await self.restoreSelectionAcceptedReplacements(true)
+            if let onFlushCompletion = self.onFlushCompletion {
+                await onFlushCompletion()
```
I would add a comment above this line (above 349 and below 348) saying "At this point, the last chunk has been received, and the session has ended, so the flush completion handler can now be invoked now that the animation is fully complete. There is no need to additionally restore the selection here since the selection would have already been restored when the session ended."

rr-codes commented on Source/WebKit/WebKitSwift/WritingTools/WKIntelligenceReplacementTextEffectCoordinator.swift:
```diff
@@ -345,7 +345,12 @@ extension WKIntelligenceReplacementTextEffectCoordinator: PlatformIntelligenceTe
+                await onFlushCompletion()
+                self.onFlushCompletion = nil
+            } else {
+                await self.restoreSelectionAcceptedReplacements(true)
```
I would add a comment above this line (above 352 and below 351) saying "At this point, the last chunk has been received, and the animation is now over, but the session has yet to end. The selection should be restored assuming it has been accepted; once the session does end, the selection will at that point then be adjusted depending on the user's action."

rr-codes commented on Source/WebKit/WebKitSwift/WritingTools/WKIntelligenceReplacementTextEffectCoordinator.swift:
```diff
@@ -345,7 +345,12 @@ extension WKIntelligenceReplacementTextEffectCoordinator: PlatformIntelligenceTe
         if chunk.finished {
             self.replacementQueue.removeFirst()
-            await self.restoreSelectionAcceptedReplacements(true)
+            if let onFlushCompletion = self.onFlushCompletion {
```
nit: should add a newline between 347 and 348

lilyspiniolas committed:
Proofreading underlines may remain if Writing Tools session ends before animation finishes
https://bugs.webkit.org/show_bug.cgi?id=286083
rdar://142980582

Reviewed by Richard Robinson.

Ensure the flush completion handler is invoked when the Writing Tools
session ends before the animation completes. As a result, the
WritingToolsController now receives a 'willEnd' call, which removes
the underlines.

* Source/WebKit/WebKitSwift/WritingTools/WKIntelligenceReplacementTextEffectCoordinator.swift:
(WKIntelligenceReplacementTextEffectCoordinator.replacementEffectDidComplete(_:)):

Canonical link: https://commits.webkit.org/289073@main
